addon/mixins/legend.js

```diff
@@ -262,26 +284,121 @@ export default Ember.Mixin.create({
+      availableLegendWidth -= Math.min(availableLegendWidth, itemWidth);
     });
+    this.set('numLegendItemsByRows', numLegendItemsByRows);
+    
+    // Calculates maximum width of label in a row, before it gets truncated by label trimmer.
+    // If labelWidth < average width per label (totalAvailableWidthForLabels/label count), then do not truncate
+    // Else if labelWidth > average, then truncate it to average
+    var calcMaxLabelWidth = function(labelWidthsArray, totalAvailableWidthForLabels) {
+      // Default the max label width to average width of an item
+      var maxLabelWidth = totalAvailableWidthForLabels / labelWidthsArray.length;
+
+      // Sort label widths to exclude all the short labels during iteration
+      labelWidthsArray = _.sortBy(labelWidthsArray);
+      for (var i = 0; i < labelWidthsArray.length; i++) {
+        var curLabelWidth = labelWidthsArray[i];
+        if (curLabelWidth < maxLabelWidth) {
+          // If the label is shorter than the max labelWidth, then it shouldn't be truncated 
+          // and hence subtract short labels from remaining totalAvailableWidthForLabels.
+          totalAvailableWidthForLabels -= curLabelWidth;
+          // Distribute the remaining width equally in remaining labels and set that as max.
+          var remainingLabelCount = labelWidthsArray.length - (i + 1);
+          maxLabelWidth = totalAvailableWidthForLabels / remainingLabelCount;
+        }
+      }
+      return maxLabelWidth;
+    };
 
-    legendItems.append('text')
-          .style('text-anchor', 'start')
-          .text(function(d) { return d.label; })
-          .attr(this.get('legendLabelAttrs'))
-          .call(labelTrimmer.get('trim'));
+    // Select labels of current row (startIdx, endIdx) and truncate if greater than labelWidth
+    var truncateLabels = function(labels, startIdx, endIdx, labelWidth) {
+      var labelTrimmer = LabelTrimmer.create({
+        getLabelSize: function() {
+          return labelWidth;
+        }
+      });
+      // Select labels from current row and apply label trimmer
+      labels.filter(function(data, idx) {
+        return (idx >= startIdx && idx < endIdx);
+      }).call(labelTrimmer.get('trim'));
+    };
 
-    var totalLabelWidth = 0;
-    legend.selectAll('text').each(function() {
-      return totalLabelWidth += this.getComputedTextLength();
-    });
-    return this.set('averageLegendLabelWidth', totalLabelWidth / this.get('legendItems.length'));
+    // Select legendItems of current row (startIdx, endIdx) and calculate total row width
+    var calcLegendRowWidth = function(legendItems, startIdx, endIdx) {
+      var rowWidth = 0;
+      legendItems.filter(function(data, idx) {
+        return (idx >= startIdx && idx < endIdx);
+      }).each(function(val, col) {
+        rowWidth = (col === 0) ? 0 : (rowWidth + 2 * legendLabelPadding);
+        rowWidth += this.getBBox().width;
+      });
+      return rowWidth;
+    };
+
+    var startIdxCurRow = 0;
```

Comment:
I'm not a big fan of unnecessary abbreviations. This should be `startIndexCurrentRow`.